### PR TITLE
[workflow] Define default `__getstate__` and `__setstate__`

### LINKED ISSUE #18426: Restrict getstate/setstate to return json and use __dict__ by default for virtual actor

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
Use `__dict__` as default getstate and setstate. Validation is added to make sure `__dict__` is able to be converted to JSON.

## Related issue number
Closes #18426 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/experimental/workflow/virtual_actor_class.py [hunk 1]
```diff
@@ -129,13 +130,34 @@ def __setstate__(self, state):
...
 
 
+def __getstate(instance):
+    if hasattr(instance, "__getstate__"):
+        state = instance.__getstate__()
+    else:
+        try:
+            state = json.dumps(instance.__dict__)
+        except TypeError as e:
+            raise ValueError("The virtual actor contains field that can't be "
```
Comment (ericl) [inline python/ray/experimental/workflow/virtual_actor_class.py:140]:
```suggestion
            raise ValueError("The virtual actor contains fields that can't be "
```

Comment (ericl) [reply python/ray/experimental/workflow/virtual_actor_class.py:140]: Can you also include the actual dict contents here in the error message?

**Review by ericl — CHANGES_REQUESTED** [review]
Can you also update workflow.rst and actor.rst?

**Review by wuisawesome — APPROVED** [review]

**Comment by iycheng** [comment]
@ericl  I update the todo in documentation ticket (https://github.com/ray-project/ray/issues/18139)

**Review by ericl — APPROVED** [review]

**Pull request merged**